udf_sbm_template.c

```diff
@@ -0,0 +1,408 @@
+    }
+    fclose(file);
+#endif /* RP_HOST */
+
+    host_to_node_int(thread_ids, n_threads);
+
+    if (myid == 0) {printf("\nFinished UDF get_inlet_thread_ids.\n"); fflush(stdout);}
+
+}
+
+  /*-------------------------*/
+ /* store_faces_normals_ids */
+/*-------------------------*/
+
+DEFINE_ON_DEMAND(store_faces_normals_ids) {
+    if (myid == 0) {printf("\n\nStarted UDF store_faces_normals_id.\n"); fflush(stdout);}
+
+    int thread, i_f, d, compute_node;
+    DECLARE_MEMORY_N(face_coords, real, ND_ND);
+    DECLARE_MEMORY_N(face_areas, real, ND_ND);
+    DECLARE_MEMORY_N(face_ids, int, mnpf);
+
+#if RP_NODE
+    Domain *domain;
+    Thread *face_thread;
+    Node *node;
+    face_t face;
+    int node_number, j;
+    real centroid[ND_ND];
+    real area[ND_ND];
+#endif /* RP_NODE */
+
+#if RP_HOST
+    char file_faces_name[ ] = "faces.dat";
+    FILE *file_faces = NULL;
+#endif /* RP_HOST */
+
+    for (thread=0; thread<n_threads; thread++) {
+
+#if RP_HOST
+        if (NULLP(file_faces = fopen(file_faces_name, "w"))) {
+            Error("\nUDF-error: Unable to open %s for writing\n", file_faces_name);
+            exit(1);
+        }
+
+#if RP_2D
+        fprintf(file_faces, "%27s %27s %27s %27s      %10s\n", "x-coordinate", "y-coordinate", "x-area", "y-area",
+        "unique-ids");
+#else /* RP_2D */
+        fprintf(file_faces, "%27s %27s %27s %27s %27s %27s      %10s\n", "x-coordinate", "y-coordinate", "z-coordinate",
+        "x-area", "y-area", "z-area",  "unique-ids");
+#endif /* RP_2D */
+#endif /* RP_HOST */
+
+#if RP_NODE
+        domain = Get_Domain(1);
+        face_thread = Lookup_Thread(domain, thread_ids[thread]);
+        n_faces = THREAD_N_ELEMENTS_INT(face_thread);
+
+        ASSIGN_MEMORY_N(face_coords, n_faces, real, ND_ND);
+        ASSIGN_MEMORY_N(face_ids, n_faces, int, mnpf);
+        ASSIGN_MEMORY_N(face_areas, n_faces, real, ND_ND);
+
+        i_f = 0;
+        begin_f_loop(face, face_thread) {
+            if (i_f >= n_faces) {Error("\nUDF-error: Index %i >= array size %i.", i_f, n_faces);}
+
+            F_CENTROID(centroid, face, face_thread); /*centroid of face returned from F_CENTROID */
+            F_AREA(area,face,face_thread); /* face normal vector returned from F_AREA */
+
+            for (d = 0; d < ND_ND; d++) {
+                face_coords[d][i_f] = centroid[d];
+                face_areas[d][i_f] = area[d];
+            }
+
+            for (j = 0; j < mnpf; j++) {
+                face_ids[j][i_f] = -1;
+            }
+
+            j = 0;
+            f_node_loop(face, face_thread, node_number) {
+                node = F_NODE(face, face_thread, node_number);
+
+                if (j >= mnpf) {Error("\nUDF-error: Index %i >= array size %i.", j, mnpf);}
+                face_ids[j][i_f] = NODE_DM_ID(node);
+                j++;
+            }
+            i_f++;
+        } end_f_loop(face, face_thread);
+
+        compute_node = (I_AM_NODE_ZERO_P) ? node_host : node_zero;
```

Comment:
Maybe call this receiver_node instead of compute_node?

Reply:
I have tried to clarify by distinguishing between receiving nodes and sending nodes.